### PR TITLE
[API Compatibility] Split `compat.py` into a folder of separate func impls

### PR DESCRIPTION
**Pull request by Enigmatisms**
### PR Category
Operator Mechanism

### PR Types
Improvements

### Description
考虑到 `paddle.compat` API 数量逐渐增多，并且需要对模块的层次调用进行对齐（比如 `torch.nn.Unfold` --> `paddle.compat.nn.Unfold`） 将原有的 `python/paddle/tensor/compat.py` 进行拆分，`compat` 拆分为 `compat` 文件夹，并设置子文件夹（`nn`, `nn/functional`）进行管理。内部 API 几乎保持不变（除Unfold，pad）

API 行为没有发生任何变化，只是组织形式存在变动。注意：
- `paddle.compat.Unfold`
- `paddle.compat.pad` 
将不再受到支持，用户应该避免调用老式写法，转而调用：
- `paddle.compat.nn.Unfold`
- `paddle.compat.nn.functional.pad`

其中，`softmax` 为一个特殊的 compat 函数。由于其需要在 `paddle.tensor` 中导入，直接归于 `compat` 管理将会触发循环引用问题。故保持原有的 `paddle/tensor/compat_softmax.py`，但在 `compat` 以及 `compat.nn.functional` 模块下导入，这使得 `softmax` 虽然定义在 `compat` 文件夹之外，但仍然有如下四种调用方式：
```py
paddle.compat.softmax                 # torch.softmax
paddle.compat.nn.functional.softmax   # torch.nn.functional.softmax
paddle.Tensor.softmax                 # torch.Tensor.softmax
paddle.special.softmax                # torch.special.softmax 
```
Pcard-89620

### CODE REVIEW

File: python/paddle/compat/__init__.py [hunk 1]
```diff
@@ -11,41 +11,31 @@
...
-    min,
-    nanmedian,
-    pad,
-    slogdet,
-    sort,
-    split,
-)
-from .tensor.compat_softmax import softmax
+from . import nn
+from .medians import median, nanmedian
```
Comment (zhwesky2010) [inline python/paddle/compat/__init__.py]: 这个可以不用每个API都单独建一个文件。
paddle主目录下是按功能分文件的，比如math.py、logic.py。paddle.sparse目录下API数量少一些，按输入个数分的文件，比如unary.py。

这一系列API不会设置很多，文件可以尽量少，每个路径一两个文件。
目前可直接在comapt/\___init\___.py、compat/nn/\___init\___.py、compat/nn/functional/\___init\___.py来实现上面所有API，如果存在循环导入或其他问题，再单独建文件。（之前softmax有这个问题）

File: python/paddle/tensor/__init__.py [hunk 2]
```diff
@@ -31,7 +31,6 @@
     real,
     shape,
 )
-from .compat_softmax import softmax as softmax
```
Comment (zhwesky2010) [inline python/paddle/tensor/__init__.py:34]: `paddle.Tensor.softmax` 目前对应的是 `paddle.compat.softmax` ，不是`paddle.softmax`。

`paddle.Tensor`去bind compat下的API是有可能的，因为后者兼容性更好。

File: python/paddle/compat/nn/Unfold.py [hunk 3]
```diff
@@ -0,0 +1,113 @@
+#   Copyright (c) 2025 PaddlePaddle Authors. All Rights Reserved.
```
Comment (zhwesky2010) [inline python/paddle/compat/nn/Unfold.py]: 放compat/nn/\___init\___.py

File: python/paddle/compat/nn/functional/pad.py [hunk 4]
```diff
@@ -0,0 +1,192 @@
+#   Copyright (c) 2025 PaddlePaddle Authors. All Rights Reserved.
```
Comment (zhwesky2010) [inline python/paddle/compat/nn/functional/pad.py]: 放compat/nn/functional/\___init\___.py

File: python/paddle/compat/softmax.py [hunk 5]
```diff
@@ -17,11 +17,10 @@
 from typing import TYPE_CHECKING
 
 from paddle import _C_ops
+from paddle.base.framework import convert_np_dtype_to_dtype_
 from paddle.framework import core, in_dynamic_or_pir_mode
 from paddle.utils.decorator_utils import ForbidKeywordsIgnoreOneParamDecorator
 
-from ..base.framework import convert_np_dtype_to_dtype_
```
Comment (zhwesky2010) [inline python/paddle/compat/softmax.py:23]: 这个API有这4个API对应的问题，前3个都是完全一致，最后一个 torch.nn.functional.softmax-> paddle.compat.nn.functional.softmax。看看会不会有导入问题。

"torch.Tensor.softmax",
"torch.special.softmax",
"torch.softmax",
"torch.nn.functional.softmax",

File: python/paddle/compat/__init__.py (hunk 1 above)
Comment (Enigmatisms) [reply python/paddle/compat/__init__.py]: ok

File: python/paddle/tensor/__init__.py (hunk 2 above)
Comment (Enigmatisms) [reply python/paddle/tensor/__init__.py:34]: ok

File: python/paddle/compat/softmax.py (hunk 5 above)
Comment (Enigmatisms) [reply python/paddle/compat/softmax.py:23]: 已经修改。softmax 确实仍然存在循环导入问题，所以并没有直接放在 compat 模块下，保留其原有的实现。compat 模块的__init__ 以及 compat.nn.functional 的 __init__ 均导入了 softmax。目前以上四种导入都没问题了。

File: python/paddle/compat/nn/functional/pad.py (hunk 4 above)
Comment (Enigmatisms) [reply python/paddle/compat/nn/functional/pad.py]: done

File: python/paddle/compat/nn/Unfold.py (hunk 3 above)
Comment (Enigmatisms) [reply python/paddle/compat/nn/Unfold.py]: done

File: python/paddle/tensor/__init__.py (hunk 2 above)
Comment (Enigmatisms) [reply python/paddle/tensor/__init__.py:34]: 见 compat/softmax.py 下的comment回复。

**Review by zyfncg — APPROVED** [review]

**Review by zhwesky2010 — APPROVED** [review]
LGTM

**Review by SigureMo — APPROVED** [review]

**Review by risemeup1 — APPROVED** [review]
